File: java/src/test/java/com/pabcablan/benchmarking/BasicMatrixMultiplierBenchmarking.java
package com.pabcablan.benchmarking;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.pabcablan.matrix.BasicMatrixMultiplier;

public class BasicMatrixMultiplierBenchmarking {

    public static void main(String[] args) {
        int[] sizes = {64, 128, 256, 512, 1024};
        int runs = 5;
        String outputDirectory = args[0];
        String outputPath = outputDirectory + "/summary_java.csv";

        System.out.println("Benchmarking in progress...");

        Map<Integer, List<Double>> times = new HashMap<>();
        Map<Integer, List<Double>> memories = new HashMap<>();

        runBenchmarks(sizes, runs, times, memories, outputPath);

        printAverageResults(sizes, times, memories);

        System.out.println("\nBenchmark finished. Results saved at: " + outputPath);
    }

        private static void ensureOutputDirExists(String outputDir) {
        File dir = new File(outputDir);
        if (!dir.exists()) {
            dir.mkdirs();
        }
    }

    private static void runBenchmarks(int[] sizes, int runs, Map<Integer, List<Double>> times, Map<Integer, List<Double>> memories, String outputPath) {
        ensureOutputDirExists(new File(outputPath).getParent());
        try (FileWriter writer = new FileWriter(outputPath)) {
            writer.write("MatrixSize,Run,TimeSeconds,MemoryMB\n");

            for (int size : sizes) {
                times.put(size, new ArrayList<>());
                memories.put(size, new ArrayList<>());

                for (int run = 1; run <= runs; run++) {
                    BasicMatrixMultiplier multiplier = new BasicMatrixMultiplier();
                    double[][][] matrices = multiplier.generateMatrices(size);

                    System.gc();
                    long memoryBefore = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
                    long start = System.nanoTime();

                    double[][] result = multiplier.multiply(matrices[0], matrices[1]);

                    long end = System.nanoTime();
                    long memoryAfter = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();

                    double timeSeconds = (end - start) / 1_000_000_000.0;
                    double memoryUsedMB = (memoryAfter - memoryBefore) / (1024.0 * 1024.0);

                    timeSeconds = Math.round(timeSeconds * 100000.0) / 100000.0;
                    memoryUsedMB = Math.round(memoryUsedMB * 100000.0) / 100000.0;

                    times.get(size).add(timeSeconds);
                    memories.get(size).add(memoryUsedMB);

                    writer.write(size + "," + run + "," + timeSeconds + "," + memoryUsedMB + "\n");
                }
            }
        } catch (IOException e) {
            System.out.println("An error occurred while writing the output file.");
            e.printStackTrace();
        }
    }

    private static void printAverageResults(int[] sizes, Map<Integer, List<Double>> times, Map<Integer, List<Double>> memories) {
        System.out.println("\n===== AVERAGE RESULTS =====");
        System.out.printf("%-10s %-15s %-15s%n", "Size", "Avg Time (s)", "Avg Mem (MB)");
        for (int size : sizes) {
            double averageTime = times.get(size).stream().mapToDouble(Double::doubleValue).average().orElse(0.0);
            double averageMemory = memories.get(size).stream().mapToDouble(Double::doubleValue).average().orElse(0.0);

            averageTime = Math.round(averageTime * 100000.0) / 100000.0;
            averageMemory = Math.round(averageMemory * 100000.0) / 100000.0;

            System.out.printf("%-10d %-15.5f %-15.5f%n", size, averageTime, averageMemory);
        }
    }
}

File: python/test/benchmarking/basic_matrix_multiplier_benchmarking.py
import sys
import time
import csv
import os
import psutil
from src.matrix.basic_matrix_multiplier import BasicMatrixMultiplier

def get_process_memory_mb():
    process = psutil.Process(os.getpid())
    return process.memory_info().rss / (1024 * 1024)

def run_benchmarks(sizes, runs, csv_path):
    times = {}
    real_memories = {}

    with open(csv_path, "w", newline='') as csvfile:
        writer = csv.writer(csvfile)
        writer.writerow(["MatrixSize", "Run", "TimeSeconds", "RealMemoryMB"])

        for size in sizes:
            times[size] = []
            real_memories[size] = []
            for run in range(1, runs+1):

                multiplier = BasicMatrixMultiplier()
                A, B = multiplier.generate_matrices(size)

                mem_before = get_process_memory_mb()
                start = time.time()
                result = multiplier.multiply(A, B)
                end = time.time()
                mem_after = get_process_memory_mb()

                time_seconds = round(end - start, 5)
                real_memory_mb = round(max(mem_before, mem_after), 5)

                times[size].append(time_seconds)
                real_memories[size].append(real_memory_mb)

                writer.writerow([size, run, time_seconds, real_memory_mb])
            avg_mem = round(sum(real_memories[size])/len(real_memories[size]), 5)

def print_average_results(sizes, times, real_memories):
    print("\n===== AVERAGE RESULTS =====")
    print("{:<10} {:<15} {:<25}".format("Size", "Avg Time (s)", "Avg Real Mem (MB)"))
    for size in sizes:
        avg_time = round(sum(times[size]) / len(times[size]), 5)
        avg_real_mem = round(sum(real_memories[size]) / len(real_memories[size]), 5)
        print("{:<10} {:<15.5f} {:<25.5f}".format(size, avg_time, avg_real_mem))

if __name__ == "__main__":
    sizes = [64, 128, 256, 512, 1024]
    runs = 5
    output_directory = sys.argv[1]
    os.makedirs(output_directory, exist_ok=True)
    csv_path = os.path.join(output_directory, "summary_python.csv")

    print("Benchmarking in progress...")

    run_benchmarks(sizes, runs, csv_path)

    times = {size: [] for size in sizes}
    real_memories = {size: [] for size in sizes}
    with open(csv_path, "r") as csvfile:
        reader = csv.DictReader(csvfile)
        for row in reader:
            size = int(row["MatrixSize"])
            times[size].append(float(row["TimeSeconds"]))
            real_memories[size].append(float(row["RealMemoryMB"]))

    print_average_results(sizes, times, real_memories)

    print("\nBenchmark finished. Results saved at:", csv_path)
